core/shared/src/main/scala/zio/ZQueue.scala

```diff
@@ -484,104 +483,118 @@ object ZQueue {
 
     val capacity: Int = queue.capacity
 
-    def offer(a: A): UIO[Boolean] = offerAll(List(a))
+    def offer(a: A): UIO[Boolean] =
+      UIO.effectSuspendTotal {
+        if (shutdownFlag.get) ZIO.interrupt
+        else {
+          val noRemaining =
+            unsafePollN(takers, 1) match {
+              case taker :: _ =>
+                unsafeCompletePromise(taker, a)
+                true
+              case Nil => false
+            }
+
+          if (noRemaining) IO.succeedNow(true)
+          else {
+            // not enough takers, offer to the queue
+            val surplus = unsafeOfferAll(queue, List(a))
+            unsafeCompleteTakers()
+
+            if (surplus.isEmpty) IO.succeedNow(true)
+            else
+              strategy.handleSurplus(surplus, queue, shutdownFlag) <* IO.effectTotal(unsafeCompleteTakers())
```

Comment:
This could be made faster by pushing the `unsafeCompleteTakers()` into `handleSurplus`, perhaps. But you have 2 `FlatMap` from `<*` and another `effectTotal` wrapper around `unsafeCompleteTakers()`.